### PR TITLE
Testing package dependencies with Degraph

### PR DESCRIPTION
**Pull request by schauder**
Adds a test that tests for dependency cycles on package an module level.

Modules in this sense are all classes that start with the same package: org.junit.gen5.x

In order to make the test pass I had to break to dependencies from test to the JUnit5TestEngine. Since the engine wasn't really part of the test, I replaced it with a mock.

I agree to https://github.com/junit-team/junit-lambda/blob/master/CONTRIBUTING.md

### CODE REVIEW

File: junit-tests/build.gradle [hunk 1]
```diff
@@ -21,9 +21,9 @@ dependencies {
 	testCompile("org.assertj:assertj-core:${assertJVersion}")
 	testCompile("junit:junit:${junit4Version}")
 	testCompile("org.mockito:mockito-core:${mockitoVersion}")
+	testCompile("de.schauderhaft.degraph:degraph-check:${degraphVersion}")
 
 	testRuntime("org.apache.logging.log4j:log4j-core:${log4JVersion}")
-	testRuntime("org.apache.logging.log4j:log4j-jul:${log4JVersion}")
```
Comment (marcphilipp) [inline junit-tests/build.gradle]: Why did you remove that?

**Comment by marcphilipp** [comment]
@schauder Thanks for the PR! Just one minor question in the diff.

File: junit-tests/src/test/java/org/junit/gen5/commons/util/org/junit/gen5/meta/DependencyTests.java [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+ * Copyright 2015 the original author or authors.
+ *
+ * All rights reserved. This program and the accompanying materials are
+ * made available under the terms of the Eclipse Public License v1.0 which
+ * accompanies this distribution and is available at
+ *
+ * http://www.eclipse.org/legal/epl-v10.html
+ */
+
+package org.junit.gen5.commons.util.org.junit.gen5.meta;
```
Comment (marcphilipp) [inline junit-tests/src/test/java/org/junit/gen5/commons/util/org/junit/gen5/meta/DependencyTests.java]: Should be just `org.junit.gen5.meta`, right?

File: junit-tests/build.gradle (hunk 1 above)
Comment (schauder) [inline junit-tests/build.gradle]: Ooops, by accident. I'll put it back in.

**Comment by marcphilipp** [comment]
@schauder Is there a way to tell degraph that `org.junit.gen5.engine.junit4` (junit4-engine), `org.junit.gen5.engine.junit5` (junit5-engine) and everything else in `org.junit.gen5.engine.**` (engine-api) are separate modules? I tried using `NamedPattern` but couldn't get it to work.

**Comment by jlink** [comment]
For legal reasons you'll have to add a paragraph to the pull request description that states that you agree to the terms in https://github.com/junit-team/junit-lambda/blob/master/CONTRIBUTING.md

**Comment by schauder** [comment]
@marcphilipp it is possible and NamedPatterns are the way to go (although you could achieve it without them) You probably used the parameters in the wrong oder. No idea why I put them the way they are ...

**Comment by schauder** [comment]
@jlink done. Pulling out org.junit.Assert right now and replacing it with import static org.hamcrest.MatcherAssert.assertThat;

**Comment by marcphilipp** [comment]
@schauder Thanks! I've merged this pull request manually. Great to finally have this check in place. :+1:

**Comment by jlink** [comment]
I'd like to pull `DependencyTests` out of junit-tests since it takes 13 secs and slows down the suite.
Any counter arguments? Pull out to a new project?

**Comment by sbrannen** [comment]
I agree with @jlink: `DependencyTests` should not be a part of the standard _pre-commit_ build. In fact, I think it shouldn't be part of the standard CI build either. Rather, I would prefer that it be part of a _nightly_ build.

**Comment by schauder** [comment]
While I agree that it is to slow for the first build step of a CI pipeline, I prefer to let it run after every push, but in a separate build step.

Another alternative would be to include such a test in each project. Since it would only run on the classes (not the jar files) it should be much faster, although still not as fast as one would wish for a proper Unit test.

**Comment by jlink** [comment]
I don't really mind to have it as a precommit step or in the ci build. But I also want a _very fast_ test suite to run (more or less) continuously.

@schauder Did you think about externalizing the use of the tool into a gradle/maven plugin that gets a DSL like configuration block? This would give users the freedom to use it in the way best fit for their project setup.

**Comment by schauder** [comment]
@jlink Yes: https://github.com/schauder/degraph/issues/62
